### PR TITLE
WIP: Initial commit

### PR DESCRIPTION
**Pull request by deanp3it**
Initail commit with module, doesn't have inheritance model or testing included

### CODE REVIEW

**Comment by lorengordon** [comment]
@mergifyio refresh

File: modules/resource-discovery/main.tf [hunk 1]
```diff
@@ -0,0 +1,7 @@
+resource "aws_vpc_ipam_resource_discovery" "this" {
+  description = var.description
+  operating_regions {
```
Comment (lorengordon) [inline modules/resource-discovery/main.tf:3]: This block can almost certainly be specified multiple times to enable multiple regions. To model that, we can use a list of `operating_regions` as the input variable, and the [dynamic block feature](https://developer.hashicorp.com/terraform/language/expressions/dynamic-blocks).

File: modules/scope/variables.tf [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+variable "description" {
+  description = "(Optional) A description for the scope you're creating."
+  type        = string
+  default     = "My IPAM Scope"
+}
+
+variable "tags" {
+  description = "(Optional) Key-value mapping of resource tags."
+  type        = map(string)
+  default     = { "Name" = "test" }
```
Comment (lorengordon) [inline modules/scope/variables.tf:15]: Defaults for tags should just about _always_ simply be an empty map

```suggestion
  default     = {}
```

File: modules/scope/variables.tf [hunk 3]
```diff
@@ -0,0 +1,16 @@
+variable "ipam_id" {
+  description = "(Required) The ID of the IPAM for which you're creating this scope."
+  type        = string
+}
+
+variable "description" {
+  description = "(Optional) A description for the scope you're creating."
+  type        = string
+  default     = "My IPAM Scope"
```
Comment (lorengordon) [inline modules/scope/variables.tf:9]: Defaults for descriptions can usually just be `null`

```suggestion
  default     = null
```

File: variables.tf [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    description = optional(string)
+    region_name = optional(string)
+    tags        = optional(map(string))
+    cascade     = optional(bool)
+  })
+  default  = {}
+  nullable = false
+
+  validation {
+    condition     = try(var.ipam.description, null) != null
```
Comment (lorengordon) [inline variables.tf:13]: This pattern feels a little awkward. Instead, just remove the "optional" marker to make description required. E.g.

```
    description = string
```

Remove this pattern everywhere for making an attribute required, and just use the builtin features of the type instead.

File: CHANGELOG.md
Comment (lorengordon) [inline CHANGELOG.md:1]: This file needs to be updated for this project

**Review by lorengordon — COMMENTED** [review]
Ok, leaving a few comments on general guidance. Not sure if you want to address these as part of this PR, or as a separate task?

File: modules/scope/variables.tf (hunk 2 above)
Comment (deanp3it) [reply modules/scope/variables.tf:15]: Updated all tags in module variables

**Comment by deanp3it** [comment]
@Mergifyio refresh

**Pull request merged**